ipa-core/src/protocol/ipa_prf/validation_protocol/proof_generation.rs

```diff
@@ -0,0 +1,517 @@
+        // RECURSION_FACTOR many points
+        // since they are arrays of fixed length, i.e. RECURSION_FACTOR,
+        // (which are filled with `F::ZERO` when there are not enough elements).
+        let r_minus_one = SmallProofGenerator::RECURSION_FACTOR - 1;
+        uv_values = loop {
+            let stop = uv_values.len() <= r_minus_one;
+
+            // generate next proof
+            let (uv_values_new, proof_from_left, prover_left_proof) =
+                SmallProofGenerator::compute_next_proof(
+                    ctx,
+                    &mut record_counter,
+                    &lagrange_table,
+                    uv_values.iter(),
+                );
+            // collect proof
+            proofs_from_left.push(proof_from_left);
+            prover_left_proofs.push(prover_left_proof);
+
+            uv_values = uv_values_new;
+
+            if stop {
+                break uv_values;
+            }
+        };
+
+        // generate masks
+        // verifier on the right has p,
+        // therefore the right share is "implicitly sent" to the right ("communicated" via PRSS)
+        let (p_mask_from_left, p_mask_prover): (Fp61BitPrime, Fp61BitPrime) =
+            ctx.prss().generate_fields(record_counter);
+        record_counter += 1;
+        // and verifier on the left has q
+        // therefore the left share is "implicitly sent" to the left (communication via PRSS)
+        let (q_mask_prover, q_mask_from_right): (Fp61BitPrime, Fp61BitPrime) =
+            ctx.prss().generate_fields(record_counter);
+        record_counter += 1;
+
+        // compute final uv values
+        let (u_values, v_values) = &mut uv_values[0];
+        // check that last elements are really Zero
+        debug_assert_eq!(
+            u_values[SmallProofGenerator::RECURSION_FACTOR - 1],
+            Fp61BitPrime::ZERO
+        );
+        debug_assert_eq!(
+            v_values[SmallProofGenerator::RECURSION_FACTOR - 1],
+            Fp61BitPrime::ZERO
+        );
+        // shift first element to last position
+        u_values[SmallProofGenerator::RECURSION_FACTOR - 1] = u_values[0];
+        v_values[SmallProofGenerator::RECURSION_FACTOR - 1] = v_values[0];
+        // set masks in first position
+        u_values[0] = p_mask_prover;
+        v_values[0] = q_mask_prover;
+
+        // generate last proof
+        let (_, proof_from_left, prover_left_proof): (
+            UVValues<Fp61BitPrime, { SmallProofGenerator::RECURSION_FACTOR }>,
+            [Fp61BitPrime; SmallProofGenerator::PROOF_LENGTH],
+            [Fp61BitPrime; SmallProofGenerator::PROOF_LENGTH],
+        ) = SmallProofGenerator::compute_next_proof(
+            ctx,
+            &mut record_counter,
+            &lagrange_table,
+            uv_values.iter(),
+        );
+
+        // collect proof
+        proofs_from_left.push(proof_from_left);
+        prover_left_proofs.push(prover_left_proof);
+
```

Comment:
This is also basically duplicate code. See lines 236 - 245. We need to find a way to avoid this duplication.